refactor: Use std::string for thread and index names

stickies-v opened the pull request:
As a follow-up to https://github.com/bitcoin/bitcoin/pull/25967#discussion_r959637189, this PR changes the return type of [`BaseIndex::GetName()`](https://github.com/bitcoin/bitcoin/blob/fa5c224d444802dabec5841009e029b9754c92f1/src/index/base.h#L120) to `const std::string&` instead of `const char*`. The first commit is not essential for this change, but since the code is touched and index names are commonly used to specify thread names, I've made the same update there.

No behaviour change, just refactoring to further phase out C-style strings.

Note: `util::ThreadRename()` used to take an rvalue ref, but since it then passes this to `SetInternalName()` by value, I don't think there's any benefit to having both an rvalue and lvalue ref function so I just changed it into lvalue ref. Not 100% sure I'm missing something?

stickies-v committed:
refactor: use std::string for thread names

stickies-v committed:
refactor: use std::string for index names

jonatack commented on src/index/base.h:
```diff
@@ -5,6 +5,8 @@
 #ifndef BITCOIN_INDEX_BASE_H
 #define BITCOIN_INDEX_BASE_H
 
+#include <string>
```
We place system header includes after ones to our codebase.

stickies-v replied:
Are there any written guidelines on this? I couldn't find anything in the developer notes and noticed both conventions being used, e.g. `blockfilter.h` for the opposite. Browsing through some more header files, it seems _most_indeed include system headers last, so I'll update it.

willcl-ark replied:
Could be a good addition to https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization if it's generally-agreed practice

stickies-v replied:
Agreed, would be a good addition. Updated all 4 index header files to import system headers last.

jonatack commented on src/index/coinstatsindex.h:
```diff
@@ -49,7 +51,11 @@ class CoinStatsIndex final : public BaseIndex
+    {
+        static std::string name{"coinstatsindex"};
+        return name;
+    }
```
Can the temporary here (and in txindex.h) be omitted? (also, why `static`)

stickies-v replied:
We need to return a `const std::string&` reference, so the underlying object can't be destroyed after the function returns, which is achieved with the `static`. I don't know of a better way to implement this but I'm open to suggestions?

jonatack replied:
Ah, right.

stickies-v replied:
No strong feelings, but I'm not sure it's an improvement:
- doesn't reduce the amount of code (disregarding newlines)
- moves the name further away from the only place where it's used, so imo slightly harder to read
- requires index_name to be initialized for every instance of the class instead of just once

So with that in mind, I think I'd rather keep it as is?

MarcoFalke replied:
It may also be possible to return a string_view (from a lifetime perspective this is no different than a reference), which works for strings and char-string literals.

string_view should work with tinyformat as well as std::string::operator+

jonatack replied:
<details><summary>Another approach :smile: </summary><p>

```diff
diff --git a/src/index/coinstatsindex.cpp b/src/index/coinstatsindex.cpp
index b9029e946a..a554cd4e3b 100644
--- a/src/index/coinstatsindex.cpp
+++ b/src/index/coinstatsindex.cpp
@@ -23,6 +23,7 @@ using node::UndoReadFromDisk;
 static constexpr uint8_t DB_MUHASH{'M'};
+static const std::string index_name{"coinstatsindex"};
 
@@ -392,6 +393,8 @@ bool CoinStatsIndex::CustomInit(const std::optional<interfaces::BlockKey>& block
 
+const std::string& CoinStatsIndex::GetName() const { return index_name; }
+
 bool CoinStatsIndex::CustomCommit(CDBBatch& batch)
diff --git a/src/index/coinstatsindex.h b/src/index/coinstatsindex.h
index 91a30fb9d9..54fff61e53 100644
--- a/src/index/coinstatsindex.h
+++ b/src/index/coinstatsindex.h
@@ -51,11 +51,7 @@ protected:
 
-    const std::string& GetName() const override
-    {
-        static std::string name{"coinstatsindex"};
-        return name;
-    }
+    const std::string& GetName() const override;
 
 public:
     // Constructs the index, which becomes available to be queried.
diff --git a/src/index/txindex.cpp b/src/index/txindex.cpp
index b719aface8..489bcd53c0 100644
--- a/src/index/txindex.cpp
+++ b/src/index/txindex.cpp
@@ -12,6 +12,7 @@
 constexpr uint8_t DB_TXINDEX{'t'};
+static const std::string index_name{"txindex"};
  
@@ -54,6 +55,8 @@ TxIndex::TxIndex(std::unique_ptr<interfaces::Chain> chain, size_t n_cache_size,
 
+const std::string& TxIndex::GetName() const { return index_name; }
+
 bool TxIndex::CustomAppend(const interfaces::BlockInfo& block)
 {
     // Exclude genesis block transaction because outputs are not spendable.
diff --git a/src/index/txindex.h b/src/index/txindex.h
index cf0ecf498f..c54ad562d8 100644
--- a/src/index/txindex.h
+++ b/src/index/txindex.h
@@ -29,11 +29,7 @@ protected:
 
-    const std::string& GetName() const override
-    {
-        static std::string name{"txindex"};
-        return name;
-    }
+    const std::string& GetName() const override;
 
 public:
 ``
</p></details>

MarcoFalke replied:
> Edit: to clarify, for example in util::ThreadRename, when passed a string_view you wouldn't be able to concatenate "b-" + name, would have to be "b-" + name.data() 

`"b-" + name.data()` is undefined behaviour if it compiles.

I was thinking that you'd be able to write `"b-"s + name`, but I was wrong.

stickies-v replied:
Thank you for the input everyone, much appreciated!

I've incorporated [sipa's suggestion](https://github.com/bitcoin/bitcoin/pull/25971#issuecomment-1234301843) to move the name logic to `BaseIndex`, which I think is an elegant solution and should hopefully also satisfy jonatack's concerns? The only downside to this approach I see is that there is some [code duplication](https://github.com/bitcoin/bitcoin/compare/4f7a2b64d8398238fc364fee9f5f2bcc4832db77..a6927e89937fdaedbc42a243dea4c90b24519378#diff-bee14d2136fca480f898cfe34d606ffb7c7460888bfb7654c575c39a77ca00ddR100-R104) in the BlockFilterIndex constructor as we need to derive the filter name from the filter type, but it's minimal enough imo.

I've also updated the thread name parameters to `string_view` wherever it made sense.

> so you could write...

Thanks for the clarification! The call to `SetInternalName()` on the next line also requires a string, so I think it's best to just expect a `string` type as parameter (reverting it to what it was before).

---

I _think_ this addresses everyone's concerns voiced here, but I'll leave the comment open for a bit longer so people can react if I misunderstood anything.

MarcoFalke replied:
Closing this thread as resolved

theStack commented:
Concept ACK

brunoerg commented:
Concept ACK

stickies-v commented:
Force pushed to [address review comments](https://github.com/bitcoin/bitcoin/pull/25971#discussion_r960899235) about simplifying the `GetName()` logic and using `string_view` instead of `const std::string&`

davidgumberg commented on src/index/blockfilterindex.h:
```diff
@@ -12,6 +12,8 @@
 #include <index/base.h>
 #include <util/hasher.h>
 
+#include <string>
```
Isn't it the case that we don't need the string include here, or in `txindex.h` or `coinstatsindex.h`?

Ellajoke replied:
Okay

hebasto commented:
Concept ACK.

stickies-v commented on src/util/thread.cpp:
```diff
@@ -10,10 +10,11 @@
+#include <string>
 
-void util::TraceThread(const char* thread_name, std::function<void()> thread_func)
+void util::TraceThread(std::string_view thread_name, std::function<void()> thread_func)
```
I agree, thanks, good catch. Made it a `const std::string` instead, and since we're already copying here I re-reverted the logic for `ThreadRename()` to use an lvalue ref again.

stickies-v replied:
Do you mean for `util::TraceThread()`? I think we explicitly want a copy here to not have to worry about lifetime? The copying I mentioned was further downstream for when we pass `thread_name` to util::ThreadRename()`, where indeed a `const std::string&` is used to avoid copying again.

MarcoFalke replied:
`TraceThread` does not return before the end of the thread, so the reference is implicitly stored as a bound argument for the whole duration of the thread, no?

Tetita replied:
> You probably want a `const std::string&` to avoid a copy.

.

stickies-v replied:
@MarcoFalke if you agree I'll revert to `string_view` again, I'm not feeling very confident here but what sipa says makes sense to me. At the very least, it shouldn't introduce any _new_ lifetime issues compared to what we had before with `const char*`?

MarcoFalke replied:
Oh, I see. The arguments to `std::thread` are copied, unless they are passed as `std::ref` ( https://en.cppreference.com/w/cpp/thread/thread/thread )

So `const std::string&` is indeed fine here, as it is a reference to the copy that was created by the `std::thread` constructor. (However, `std::string` should be fine as well and come with only the overhead equivalent to constructing an empty `std::string`)

MarcoFalke replied:
For reference, the following gives an `AddressSanitizer: stack-use-after-scope`, because the argument was passed with `std::ref`, but does not if the argument is not passed with `std::ref`.

<details><summary>`std::thread`/`std::ref` test</summary>

```cpp
#include <chrono>
#include <iostream>
#include <thread>

void fun(const std::string &a) {
  std::this_thread::sleep_for(std::chrono::milliseconds{10});
  std::cout << a << std::endl;
}

int main() {
  std::thread a;
  {
    std::string s{"s"};
    a = std::thread{fun, std::ref(s)};
  }
  a.join();
}
```

</details>

MarcoFalke replied:
To clarify, anything is fine here, but `string_view` can lead to lifetime issues if there is a leftover caller that uses c_str on a temporary string. For example, this is UB:


```cpp
#include <chrono>
#include <iostream>
#include <thread>

void fun(std::string_view a) {
  std::this_thread::sleep_for(std::chrono::milliseconds{10});
  std::cout << a << std::endl;
}

int main() {
  std::thread a;
  {
    std::string s{"s"};
    a = std::thread{fun, s.c_str()};
  }
  a.join();
}

stickies-v replied:
Thanks for the background info MarcoFalke, that's really helpful.

> can lead to lifetime issues if there is a leftover caller that uses c_str on a temporary string

Would you agree that that would just be bad implementation, and not something we should expect downstream functions to try and fix?

MarcoFalke replied:
Previously passing a `std::string` was only possible by calling `c_str()` on it, so this UB may persist in a very unlikely and unfortunate silent merge conflict. (Changing `const char*` to `std::string_view` shouldn't introduce any new UB, but it wouldn't fix it either, if it existed previously). Though, again that seems unlikely so I think anything is fine here and this is almost a style question.

hebasto replied:
FWIW, in our code base, all `util::TraceThread` callers pass either a string literal (in most cases) or a `BaseIndex::GetName()` return value of `const std::string&` type (the only case).

stickies-v reviewed (commented):
Force pushed to address review feedback regarding:
- `util::TraceThread()` to take a `std::string` copy and avoid potential issues with lifetime introduced by the threading, also updated `util::ThreadRename()` to take a const lvalue ref to avoid copying twice
- remove unused includes leftover from earlier versions

stickies-v commented:
Force pushed to partly reverse the previous force push based on [sipa's suggestion](https://github.com/bitcoin/bitcoin/pull/25971#discussion_r961699046), again using a `string_view` for `util::TraceThread()` and undoing the changes to `util::ThreadRename()` (now no longer in this PR's diff).

My apologies to reviewers for the back and forth.

stickies-v commented:
Force pushed to rebase on latest master to fix failing CI (wallet_groups.py) - no other changes.

MarcoFalke commented:
My preference would still be to use `std::string` in `TraceThread` (https://github.com/bitcoin/bitcoin/pull/25971#discussion_r960917554) as a belt-and-suspenders, but the current code is not worse than it was previously.

review ACK e5d9c6f0571ee182f34e0d82c20e5ca2da30ad85 🎯

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK e5d9c6f0571ee182f34e0d82c20e5ca2da30ad85 🎯
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgoCQv+P5w/6IGCCYaSa+M7unM9uCN6fxnRgqaJlASlzu/rGzN0VJcdS404NTdL
FpIpXE2j+kwhgAc6eM+87byb2lR7sNgL/CAC7v1M8tXNMpTMJwlwgLsCykFY7iE9
pK6u9BocyloYAcDNw4u7v2Mph7KTxUIXxqZhz9fFR2dPJrxc6eTmP8OXaRiYlbAo
9BxAw79cQY3+2F3V0THByfcgQdyZO7BF2gpZCNgrY84bwIbotsTgtGU9qLpv46V9
2GBgy7v5jUqsFYPVtnM/cs1NVou779N/rCDH2+4uDBiXyYf6VM8d9ZEYIkzayPmC
hclDiXMwfpZ0hVnbdBhpmYGp9Vb4UoMkZiLleJdlJQsp3R1e3RBCDGPDsy1iV8Aw
mQMfOkiDQxC91eyizLv3eJ8RnUc7vwE9W+MW6FKQzw1O8vaFKfGJEoIK/nb2aMXr
Q92DNYhFxzZpQHG5MfoJj20ySJP/woq7xG/7xpH/zW4QuSaxW6OvqZ3GZFUNROGc
zT0TuUiQ
=b37U
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented:
re-ACK 200d84d5681918523d982b9ddf60d1127edcb448 only change is rebase 🚼

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 200d84d5681918523d982b9ddf60d1127edcb448 only change is rebase 🚼
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgPNQwAw5LrCWUbRWxPIUc1OLrGT75Vv9alSNady748D38GlxcFc3CBahCxmU4Q
sf3vpVbHN0Mo8Y06vNB0mBZ1r4lcL5U73rGzDELCNFEixh08PTELLrcdWCFLk/Dp
ScpDdGRyMx4fRkjwaYFbUZsq4qUpfPjYTQBsHKyOn/J9RR/yzbMZOP0IGtonG10t
jmO52IAPgcD2duZRCY7HZSsJtAmnaFmV0geDH1LdL0j5G2y76Sqxzid/0H6KeoaK
Tp7lPcf77T/o8lNyYllYAYtEIQp7Gg9yvQOYvmkuAKxyx9f5JOEQtw936zCt5BXt
DgHD5uve0WnR0vp1v1sowAPci/l8D9fVh9iiWQG8oqHJ8Iha4YLaXicsBrJsEbav
9BR+bIdvUw99Bg8UK1XQdEwczVFYrExhzbP50W04DM3HpD3kjlwdZhvwON6P5XKt
2O8A7wFT+KO4qjno+8DItg1FKeVh9kRuMZ7ptXq4RJVyob3L/jT8gTvKgVkcxj5L
N0aIxOMu
=fh6y
-----END PGP SIGNATURE-----
```


</details>

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25971/commits/200d84d5681918523d982b9ddf60d1127edcb448

stickies-v committed:
scripted-diff: rename pszThread to thread_name

Since it is now a string_view instead of a const char*, update the
name to reflect that the variable is no longer a "Pointer to
String, Zero-terminated" (psz).

-BEGIN VERIFY SCRIPT-
sed -i s/pszThread/thread_name/ $(git grep -l pszThread src)
-END VERIFY SCRIPT-

stickies-v commented on src/util/system.cpp:
```diff
@@ -831,7 +831,7 @@ std::string HelpMessageOpt(const std::string &option, const std::string &message
 }
 
-static std::string FormatException(const std::exception* pex, const char* pszThread)
+static std::string FormatException(const std::exception* pex, std::string_view pszThread)
```
Oh that's good to know, thanks. I've added 26cf9ea to fix that with a scripted-diff.

MarcoFalke commented:
review ACK 26cf9ea8e44d7fd6450336f567afaedd1275baf7 only change is new scripted-diff 😀

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 26cf9ea8e44d7fd6450336f567afaedd1275baf7 only change is new scripted-diff 😀
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiQUAwAuj4HULyQMBta0XNIDl0zy4CCXiT1ZBVXSQoMv0mBjQPP7l5m45FhtTQe
ThS7P6W0rwQq2jGwmPjucd29gS5LuTuf5tBIoExY17u/TDgKa53wz37ggP5d574V
PypsOaugnlpipMrlNBiOguxNS9KjE9n4MrXVF67Pnm+SbcqIW9aOlpLODNYO7MHd
/T+0wy03Vtmknsze7e4v9xp8HxQ2kZab11auPKSMVfuACDhJEBul9a/xrlDILjTG
QvfCG2BRdIaUzMLgtLWGTOD8Ilq7h3ObKjnKW4aH42QIRZi/Wfp3tQhvKkOZaW7N
zAISfPqIk1ZelkAcEPjerE62fdPnWRRBHuZGv4+Vrk0eBgnt422PuvohTaUA98r8
mOL9zcXRXT03VNoV2SsHFBcs2tPQ6K74O7CtB5PZWPWJVvb4HcJt0n9PMUuba5HS
fUlnI5OB6PwLgrFiABiR+fmbskUSdcv+CY0ZT4IDYGB2/m2egzo20HKBWYwVBEO3
T5vJv1+V
=hJ1B
-----END PGP SIGNATURE-----
```


</details>

hebasto commented on src/util/thread.cpp:
```diff
@@ -10,10 +10,12 @@
 #include <exception>
 #include <functional>
+#include <string>
+#include <utility>
```
nit: Why this header?

MarcoFalke replied:
std::move

hebasto replied:
```
$ git grep move -- src//util/thread.cpp | wc -l
0
```

MarcoFalke replied:
iwyu:

```
util/thread.h should add these lines:
#include <string_view>  // for string_view

util/thread.h should remove these lines:
- #include <string>  // lines 9-9

The full include-list for util/thread.h:
#include <functional>   // for function
#include <string_view>  // for string_view
---

util/thread.cpp should add these lines:

util/thread.cpp should remove these lines:
- #include <utility>  // lines 14-14

The full include-list for util/thread.cpp:
#include <util/thread.h>
#include <logging.h>           // for LogPrintf_, LogPrintf
#include <util/system.h>       // for PrintExceptionContinue
#include <util/threadnames.h>  // for ThreadRename
#include <exception>           // for exception
#include <functional>          // for function
#include <string>              // for allocator, string
---

hebasto reviewed (approved):
ACK 26cf9ea8e44d7fd6450336f567afaedd1275baf7, I have reviewed the code and it looks OK.

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/25971/commits/26cf9ea8e44d7fd6450336f567afaedd1275baf7
